### PR TITLE
Fix latent/caption mismatch in SD3 dreambooth lora trainer with cache_latents + shuffle

### LINKED ISSUE #14430: [SD3 DreamBooth] Cached latents can be paired with the wrong per-image captions when shuffle=True
### Describe the bug

The SD3 DreamBooth LoRA trainer creates the training dataloader with shuffle enabled.
When latent caching is enabled, the script traverses the shuffled dataloader once and stores the VAE latent distributions in a list. Training then starts another traversal of the same shuffled dataloader, but retrieves cached latents using only the current batch position.
Because the second traversal can produce a different sample order, a per-image caption in the current batch may be paired with the cached latent of another image. This problem is mostly hidden when every image uses the same instance prompt, but it can silently corrupt training when the caption column contains a different caption for each image.
Cached latents should be stored and retrieved using a stable dataset index or sample ID instead of the dataloader step. This would preserve shuffling while keeping each image, latent and caption correctly associated.
A similar correspondence issue was reported for the Qwen-Image trainer in issue #12124.

### Reproduction

The issue follows directly from the current data flow in train_dreambooth_lora_sd3.py.
The training dataloader is created with shuffle enabled. When latent caching is enabled, the first traversal appends each latent distribution to a list. Training starts a second traversal of the shuffled dataloader and reads the cached latent using latents_cache[step].
The first and second traversals are not guaranteed to have the same sample order. Therefore, the current per-image caption and latents_cache[step] are not guaranteed to belong to the same dataset sample.
This can be observed by printing the sample index during the latent caching traversal and again during the first training epoch.

### Logs

```shell
No runtime exception is raised. This is a silent image-caption correspondence issue that can affect the learned result without producing an error message.
```

### System Info

Diffusers: 0.40.0.dev0
Diffusers source commit: d6726f38a0c5ca6c06a8f227fb7bade3486ed98d
Python: 3.12
PyTorch: 2.8.0
CUDA: 12.8
Platform: Ubuntu 22.04
GPU: NVIDIA GeForce RTX 4090, 24GB
Training script: examples/dreambooth/train_dreambooth_lora_sd3.py
Model: stabilityai/stable-diffusion-3.5-medium
Dataset mode: local ImageFolder with per-image captions
Latent caching: enabled
Dataloader shuffling: enabled

### Who can help?

@sayakpaul @DN6

### PR DESCRIPTION
**Pull request by lcheng321**
Fixes #14430

## Root cause

Latents were cached keyed by dataloader step, but the dataloader has `shuffle=True` and is traversed twice: once during caching, once during training. Each traversal produces a different sample order. `latents_cache[step]` returns whatever image happened to land at that step during caching, not the image in the current training batch. No exception is raised, the mismatch is silent.

## Fix

Cache latents keyed by stable dataset index instead of step. Index is generated in `DreamBoothDataset.__getitem__` and passed through `collate_fn`.

Also added a check that rejects `--cache_latents` combined with `--with_prior_preservation`, since class images are not indexed the same way and would hit the same bug through a different path.

## Scope

This PR only covers `train_dreambooth_lora_sd3.py`. It does not touch the Qwen-Image trainer or #12124, which is a separate file with a similar but independent issue.

### 1. Isolated repro, no model or GPU required

Dataset index used directly as the tensor value, so any mismatch between cached value and actual batch value is visible by eye. Same dataloader, same seed, two traversals.

Before fix, 4/4 batches mismatched:

step=0: latents_cache[0]=[2.0, 0.0] vs actual pixel_values=[0.0, 3.0] MISMATCH
step=1: latents_cache[1]=[5.0, 3.0] vs actual pixel_values=[7.0, 1.0] MISMATCH
step=2: latents_cache[2]=[1.0, 4.0] vs actual pixel_values=[6.0, 2.0] MISMATCH
step=3: latents_cache[3]=[7.0, 6.0] vs actual pixel_values=[5.0, 4.0] MISMATCH

After fix, 4/4 batches matched:

step=0: indices=[0, 3] cached=[0.0, 3.0] actual=[0.0, 3.0] OK
step=1: indices=[7, 1] cached=[7.0, 1.0] actual=[7.0, 1.0] OK
step=2: indices=[6, 2] cached=[6.0, 2.0] actual=[6.0, 2.0] OK
step=3: indices=[5, 4] cached=[5.0, 4.0] actual=[5.0, 4.0] OK
all_match: True

### 2. Existing test suite

pytest examples/dreambooth/test_dreambooth_lora_sd3.py examples/dreambooth/test_dreambooth_sd3.py -v

11 passed in 344.51s

All existing tests pass, including `test_dreambooth_lora_latent_caching`.

### CODE REVIEW

**Comment by lcheng321** [comment]
Hey @sayakpaul @DN6 would appreciate a review when you have a moment.
